spmd/compiler/graph_utils.py

```diff
@@ -22,52 +19,96 @@ class OP(str, Enum):
 
 
-def create_name_to_node_map(gm: fx.GraphModule) -> Dict[str, fx.Node]:
-    """utility to put graph module into a node map for easier adjustments"""
-    mapping = {}
-    for node in gm.graph.nodes:
-        mapping[node.name] = node
-    return mapping
-
-
-def get_node_tensor_numel_shape(
-    node: fx.Node,
-) -> Tuple[Optional[int], Optional[torch.Size]]:
-    """takes an fx node, and if tensor data available, optionally displays and returns numel"""
-    size = None
-    shape = None
-    tmeta: Optional[TensorMetadata] = node.meta.get("tensor_meta")
-    if tmeta is None:
-        raise RuntimeError(f"failed to locate metadata for node {node.name}")
-    size = tmeta.shape.numel()
-    shape = tmeta.shape
-
-    rank0_debug(
-        logger,
-        f"--> tensor of size {size} and shape {shape} found for {node=}\n",
-    )
-
-    return size, shape  # type: ignore
+class CommType(str, Enum):
+    ALLREDUCE = "allreduce_"
+    ALLGATHER = "allgather_"
+    BROADCAST = "broadcast_"
+    REDUCESCATTER = "reduce_scatter_"
+    SCATTER = "scatter_"
+
+
+comm_block_op_sequence: Tuple[Union[str, Set[CommType]], ...] = (
+    "clone",
+    "_tensor_constant",
+    "_tensor_constant",
+    # The supported communication type.
+    set((CommType.ALLREDUCE,)),
+    "comm_result",
+    "getitem",
+    "getitem",
+    "wait_comm",
+)
+
+
+def get_comm_block_nodes(
+    wait_node: fx.Node, comm_type: CommType
+) -> Tuple[int, List[fx.Node]]:
+    """
+    Given a wait_comm node, find out all the nodes belong to this communcation.
+
+    Args:
+        wait_node(fx.Node): The target wait_comm node.
+        comm_type(CommType): The communication type of this communication block.
+            Currently, only allreduce is supported. An exception will be raised
+            if other values are passed.
+    Returns:
+        comm_idx(int): The index to the communication node in the return list.
+        node_list(List[fx.Node]): The list that contain the nodes in the order
+           of inserting to the graph.
+    """
+    if not wait_node.name.startswith("wait_comm"):
+        raise ValueError(
+            "Passing a wait_node that name does not start with ``wait_comm``. "
+            f"Name is {wait_node.name}, OP is {wait_node.op}."
+        )
+    node = wait_node
+    node_list = []
+    for i, prefix in enumerate(reversed(comm_block_op_sequence)):
+        node_list.append(node)
+        if isinstance(prefix, set):
+            if comm_type not in prefix:
+                raise ValueError(f"Not supported CommType {comm_type}")
+            prefix = comm_type
+            comm_idx = i
+        assert node.name.startswith(
+            prefix
+        ), f"Comm block op sequence mismatches, {node.op} {node.name} {i} {prefix}."
+        node = node.prev
+
+    comm_idx = len(node_list) - comm_idx - 1
+    node_list.reverse()
+
+    return comm_idx, node_list
+
+
+def get_node_tensor_metadata(
+    node: fx.Node, must_present: bool = True
```

Comment:
minor nit - I think "required" or "is_required" is cleaner for a param name vs "must_present".  
I initially interpreted 'must_present" as meaning "must debug show", vs what you really mean is "must_be_present".